### PR TITLE
Python 3 imports: additional work on builtins, relative imports, importlib

### LINKED ISSUE #8712: Python 3/NVDAObjects.IAccessible modules: absolute import fails if a given IAccessible class implementation is located outside the main module
Hi,

A note about the title: suffice to say that pre-transition research has progressed to a point where imports and small things can be tested here and there.

### Background:
Building on #8590:

In NVDAObjects/IAccessible module, when loading a support class for a control NvDA has just encountered, it'll check class name and role inside a static map, which is then used to direct NVDA as to where the support routine can be found. In case the resulting class string is "mod.class", NVDA will use import machinery (__import__ function) to load the appropriate module.

However, as noted in #8590, because Python 3 does not allow import of the form "from foo import something" when "foo" is to be imported from a module located in the same location, a dot is placed in front of the module name. Unfortunately, NvDAObjects/IAccessible does not know that this is the case in Python 3, so it raises "ModuleNotFoundError" when trying to import an MSAA module from __init__.py (for example, when importing sysListView.32.List, it works in Python 2, but does not in Python 3).

### Steps to reproduce:
1. Create a package inside a folder with an __init__.py file and another module.
2. Try using the import of the form in python 3:

mod=__import__(modString,globals(),locals(),[])

Where "modString" is the name of the module inside the same folder as __init__.py.

### Actual behavior:
Does not work in Python 3.

### Expected behavior:
Works in Python 3.

#### NVDA Installed/portable/running from source:
N/A

#### NVDA version:
N/A

#### Windows version:
N/A

#### Name and version of other software in use when reproducing the issue:
Python 2.7.15, 3.7.0

#### Other information about your system:
N/A

#### Does the issue still occur after restarting your PC?
Not applicable

#### Have you tried any other versions of NVDA?
Not applicable

### Possible solutions:

1. Force Python to load the modules in question by manipulating sys.path.
2. Insert a dot when encountering class string of the form "mod.class" so __import__ function can import .mod instead.
3. Try using importlib (Python 3).

Thanks.

### LINKED ISSUE #8724: Python 3/Python Console and others: __builtin __ -> builtins
Hi,

Continuing from #8590: The module "__builtin__" is no more, replaced by "builtins" in Python 3. This affects Python Console.

### Steps to reproduce:
With both Python 2.7 and 3.7 interpreters running, try doing the following:

import __builtin__

### Actual behavior:
Module not found error is raised in Python 3.

### Expected behavior:
The "__built__in" module is imported.

#### NVDA Installed/portable/running from source:
N/A

#### NVDA version:
N/A

#### Windows version:
N/A

#### Name and version of other software in use when reproducing the issue:
Python 2.7, 3.7

#### Other information about your system:
Not applicable

#### Does the issue still occur after restarting your PC?
N/A

#### Have you tried any other versions of NVDA?
N/A

### Solution:
Apply the same procedure as #8590: import Python 2 version first, and then load Python 3 equivalent.

A note regarding a pull request for this issue: the resulting pull request will not only address this issue, but also import level problem in IAccessible object module, as well as plugging more relative import problems (mostly noted in UIA object module).

Thanks.

### LINKED ISSUE #8768: Python 3/imports: use importlib instead of calling __import__ function directly
Hi,

Stemming from an observation in #8712: Python 2 does come with importlib, which does have a more elegant way of loading modules via importlib.import_module. This is more readable than using __import__ function directly.

### Steps to reproduce:
Try using both __import__ and importlib.impot_module to load an external module.

### Actual behavior:
Module loads via either method.

### Expected behavior:
Module loads.

#### NVDA Installed/portable/running from source:
N/A

#### NVDA version:
N/A

#### Windows version:
N/A

#### Name and version of other software in use when reproducing the issue:
Python 2.7.15, 3.7.0

#### Other information about your system:
N/A

#### Does the issue still occur after restarting your PC?
N/A

#### Have you tried any other versions of NVDA?
N/A

### When this should be implemented:
Preferably as part of NVDA 2019.1.

Thanks.

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #8712
Fixes #8724
Fixes #8768

### Summary of the issue:
More work on Python 3 imports, including relative imports and changing __builtin__ to builtins, as well as using importlib to import modules at runtime.

### Description of how this pull request fixes the issue:
Changes made:

* Import __builtin__ -> builtins (Python Console, language handler).
* GUI and IAccessible objects: more relative imports (from . import ...).
* Use importlib to import modules at runtime, including for IAccessible modules.

### Testing performed:
Tested with Python 2.7 interpreter to make sure they are imported.

### Known issues with pull request:
None

### Change log entry:
None

### CODE REVIEW

**Comment by leonardder** [comment]
[**@josephsl**](https://github.com/josephsl) commented on [14 Sep 2018, 16:25 CEST](https://github.com/nvaccess/nvda/pull/8727#issuecomment-421375157 "2018-09-14T14:25:28Z - Replied by Github Reply Comments"):
> Hi, importlib is new in Python 3 (if I read this correctly).

It is also there in Python 2.7. 

> Help on built-in function __import__ in module __builtin__:
>
> __import__(...)
>     __import__(name, globals={}, locals={}, fromlist=[], level=-1) -> module
>    
>     Import a module. Because this function is meant for use by the Python
>    interpreter and not for general use, it is better to use
>    importlib.import_module() to programmatically import a module.

**Comment by josephsl** [comment]
Hi, I see. In this case, I propose doing this via a new pull request. Thanks.

**Comment by leonardder** [comment]
I'd rather see this as part of this one. It creates unnecessary overhead if we have to change one line of code in this pr, and then change it again in the next.

See also #8768.

**Comment by josephsl** [comment]
Hi,

By the way, there is an unused "import Queue" in synthDrivers.espeak (no underscore), so I'm ignoring that one. People who will be working on transition should note that this import causes an exception, so should be removed then.

Thanks.

File: source/appModuleHandler.py [hunk 1]
```diff
@@ -93,8 +94,7 @@ def getAppNameFromProcessID(processID,includeExt=False):
 	# Try querying the app module for the name of the app being hosted.
 	try:
 		# Python 2.x can't properly handle unicode module names, so convert them.
-		mod = __import__("appModules.%s" % appName.encode("mbcs"),
```
Comment (leonardder) [inline source/appModuleHandler.py]: This is definitely going to fail on python 3

```
>>> "hello, %s" % "test".encode("mbcs")
"hello, b'test'"
```

str.format gives the same result.

It should be sufficient to check if not isinstance(appName, str) before encoding to mbcs.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

**Comment by leonardder** [comment]
I'm again not very happy with the idea to touch code in one pr, knowing that it requires another change in a subsequent pr. then, we either better leave the code alone in this pr or fix it for good. I'm happy with either way, but not with leaving as is, honestly.

**Comment by josephsl** [comment]
Hi, I’m willing to make an exception by addressing encoding problem as part of this PR with a comment stating that Python 3 improves Unicode support. Thanks.

**Comment by josephsl** [comment]
Hi,

Addressed the import concern as of today's commit. As noted in the commit message, I'm making an exception, as nobody else calling import_module calls name.encode apart from app module handler.

Thanks.

**Comment by leonardder** [comment]
Never mind, I was horribly wrong there. This place actually encodes (goes from unicode from bytes), my head was telling me that we had to decode (from bytes to unicode). So if it would have been isinstance, the right clause would be `if not isinstance(appname, str):`

**Review by LeonarddeR — APPROVED** [review]

**Comment by josephsl** [comment]
Hi,

@feerrenrut, may I request a second look for this one? I'm interested in seeing if conflicts may arise when attempting to merge this pull request in the future.

Thanks.

**Review by feerrenrut — APPROVED** [review]
This looks good to me. Given that it is tested against python 2.7, I think we should merge it in now. We don't have any intention of continuing to support Python 2.7 once we move over to 3 so I will add a task to the python 3 project board to remove any 2.7 specific code.

**Comment by leonardder** [comment]
I agree.

**Comment by josephsl** [comment]
Possible regression reported: multi-byte file name issue. Multiple possibilities, including Python 2.7.16, this PR, or something else.

Thanks.

**Comment by leonardder** [comment]
@josephsl: Could you please file this again against threshold?

**Comment by josephsl** [comment]
Hi, yep that’ll be done (today at the earliest) with at least two separate PR’s: imports in one package, importlib in another. Thanks.

**Pull request merged**